checker/src/main/java/org/checkerframework/checker/guieffect/GuiEffectVisitor.java

```diff
@@ -322,6 +349,51 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
             }
 
             callerEffect = atypeFactory.getDeclaredEffect(callerElt);
+            final DeclaredType callerReceiverType =
+                    this.visitorState.getClassType().getUnderlyingType();
+            assert callerReceiverType != null;
+            final TypeElement callerReceiverElt = (TypeElement) callerReceiverType.asElement();
+            // Note: All these checks should be fast in the common case, but happen for every method call inside the
+            // anonymous class. Consider a cache here if profiling surfaces this as taking too long.
+            if (TypesUtils.isAnonymous(callerReceiverType)
+                    // Skip if already inferred @UI
+                    && !effStack.peek().isUI()
+                    // Ignore if explicitly annotated
+                    && !atypeFactory.fromElement(callerReceiverElt).hasAnnotation(AlwaysSafe.class)
+                    && !atypeFactory.fromElement(callerReceiverElt).hasAnnotation(UI.class)) {
+                boolean overridesPolymorphic = false;
+                Map<AnnotatedTypeMirror.AnnotatedDeclaredType, ExecutableElement>
+                        overriddenMethods =
+                                AnnotatedTypes.overriddenMethods(elements, atypeFactory, callerElt);
+                for (Map.Entry<AnnotatedTypeMirror.AnnotatedDeclaredType, ExecutableElement> pair :
+                        overriddenMethods.entrySet()) {
+                    AnnotatedTypeMirror.AnnotatedDeclaredType overriddenType = pair.getKey();
+                    AnnotatedExecutableType overriddenMethod =
+                            AnnotatedTypes.asMemberOf(
+                                    types, atypeFactory, overriddenType, pair.getValue());
+                    if (atypeFactory.getDeclAnnotation(
+                                            overriddenMethod.getElement(), PolyUIEffect.class)
+                                    != null
+                            && atypeFactory.getDeclAnnotation(
+                                            overriddenType.getUnderlyingType().asElement(),
+                                            PolyUIType.class)
+                                    != null) {
+                        overridesPolymorphic = true;
+                        break;
+                    }
+                }
+                // Perform anonymous class polymorphic effect inference:
```

Comment:
Why is effect inference in the Visitor instead of in the AnnotatedTypeFactory?

Reply:
@wmdietl The reason for this is that effect inference depends on visiting the method invocations inside the lambda body or the methods of the anonymous inner class for which inference is being performed. As far as I know, this information is only present during the Visitor traversal.

There was a lot of talk in the original lambda inference PR about using `GuiEffectTreeAnnotator`, but I couldn't get it to work, either because some necessary info was missing (`visitorState`, see [this discussion](https://github.com/typetools/checker-framework/pull/1481#discussion_r135813364)), or because certain nodes weren't being visited at all (e.g. `visitMethodInvocation` see [this discussion](https://github.com/typetools/checker-framework/pull/1481#issuecomment-326118260)). If there is a better way to do this, I would love to know of it, though.